File: src/main/java/com/camerapoints/ui/PointPanel.java
package com.camerapoints.ui;

import com.camerapoints.CameraPointsPlugin;
import com.camerapoints.models.CameraPoint;
import com.camerapoints.models.CameraPointGroup;
import net.runelite.client.config.Keybind;
import net.runelite.client.ui.ColorScheme;
import net.runelite.client.ui.FontManager;
import net.runelite.client.ui.components.FlatTextField;
import net.runelite.client.util.ImageUtil;

import javax.swing.BorderFactory;
import javax.swing.ImageIcon;
import javax.swing.JButton;
import javax.swing.JCheckBox;
import javax.swing.JComboBox;
import javax.swing.JLabel;
import javax.swing.JOptionPane;
import javax.swing.JPanel;
import javax.swing.JSpinner;
import javax.swing.JToggleButton;
import javax.swing.SpinnerNumberModel;
import javax.swing.border.CompoundBorder;
import javax.swing.border.EmptyBorder;
import javax.swing.border.LineBorder;
import javax.swing.border.MatteBorder;
import java.awt.BorderLayout;
import java.awt.Color;
import java.awt.Cursor;
import java.awt.Dimension;
import java.awt.FlowLayout;
import java.awt.event.KeyAdapter;
import java.awt.event.KeyEvent;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;

public class PointPanel extends JPanel
{
    private static final int ZOOM_LIMIT_MIN = -272;
    private static final int ZOOM_LIMIT_MAX = 1400;

	private final CameraPointsPlugin plugin;
	private final CameraPointGroup group;
	private final CameraPoint cameraPoint;
	private final Runnable reloadFunction;

	private final ImageIcon fromGameIcon = new ImageIcon(ImageUtil.loadImageResource(CameraPointsPlugin.class, "from_game_icon.png"));
	private final ImageIcon fromGameHoverIcon = new ImageIcon(ImageUtil.loadImageResource(CameraPointsPlugin.class, "from_game_icon_blue.png"));
	private final ImageIcon fromGamePressedIcon = new ImageIcon(ImageUtil.alphaOffset(fromGameHoverIcon.getImage(), -50));
	private final ImageIcon deleteIcon = new ImageIcon(ImageUtil.loadImageResource(CameraPointsPlugin.class, "delete_icon.png"));
	private final ImageIcon deleteHoverIcon = new ImageIcon(ImageUtil.loadImageResource(CameraPointsPlugin.class, "delete_icon_red.png"));
	private final ImageIcon deletePressedIcon = new ImageIcon(ImageUtil.alphaOffset(deleteHoverIcon.getImage(), -50));
    private final ImageIcon zoomIcon = new ImageIcon(ImageUtil.loadImageResource(CameraPointsPlugin.class, "zoom_icon.png"));
    private final ImageIcon zoomPressedIcon = new ImageIcon(ImageUtil.loadImageResource(CameraPointsPlugin.class, "zoom_icon_blue.png"));

	private final JCheckBox enabledBox = new JCheckBox();
	private final FlatTextField nameInput = new FlatTextField();
	private final JLabel saveLabel = new JLabel("Save");
	private final JLabel cancelLabel = new JLabel("Cancel");
	private final JLabel renameLabel = new JLabel("Rename");
	private final JButton keybindButton = new JButton();
	private final JComboBox<CameraPoint.Direction> directionDropdown = new JComboBox<>(CameraPoint.Direction.values());
	private final JToggleButton applyZoomButton = new JToggleButton();
	private final JLabel updateZoomLabel = new JLabel(fromGameIcon);
	private final JLabel deleteLabel = new JLabel(deleteIcon);
    private final JSpinner zoomSpinner = new JSpinner(new SpinnerNumberModel(ZOOM_LIMIT_MIN, ZOOM_LIMIT_MIN, ZOOM_LIMIT_MAX, 1));

	public PointPanel(CameraPointsPlugin plugin, CameraPointGroup group, CameraPoint cameraPoint, Runnable reloadFunction)
	{
		this.plugin = plugin;
		this.group = group;
		this.cameraPoint = cameraPoint;
		this.reloadFunction = reloadFunction;

		setLayout(new BorderLayout());
		setBorder(BorderFactory.createLineBorder(Color.LIGHT_GRAY));
		setBackground(ColorScheme.DARKER_GRAY_COLOR);

		JPanel namePanel = new JPanel(new BorderLayout());
		namePanel.setBackground(ColorScheme.DARKER_GRAY_COLOR);
		namePanel.setBorder(new CompoundBorder(
			new MatteBorder(0, 0, 1, 0, ColorScheme.DARK_GRAY_COLOR),
			new LineBorder(ColorScheme.DARKER_GRAY_COLOR)));

		JPanel nameActions = new JPanel(new BorderLayout(4, 0));
		nameActions.setBackground(ColorScheme.DARKER_GRAY_COLOR);
		nameActions.setBorder(new EmptyBorder(0, 4, 0, 8));

		configureNameActions(nameActions);

		nameInput.setText(cameraPoint.getName());
		nameInput.setBorder(null);
		nameInput.setEditable(false);
		nameInput.setBackground(ColorScheme.DARKER_GRAY_COLOR);
		nameInput.setPreferredSize(new Dimension(0, 24));
		nameInput.getTextField().setForeground(Color.WHITE);
		nameInput.getTextField().setBorder(new EmptyBorder(0, 8, 0, 0));
		nameInput.addKeyListener(new KeyAdapter()
		{
			@Override
			public void keyPressed(KeyEvent e)
			{
				if (e.getKeyCode() == KeyEvent.VK_ENTER)
				{
					saveName();
				}
				else if (e.getKeyCode() == KeyEvent.VK_ESCAPE)
				{
					cancelRename();
				}
			}
		});
		nameInput.getTextField().addMouseListener(new MouseAdapter()
		{
			@Override
			public void mouseClicked(MouseEvent e)
			{
				if (!Helper.checkClick(e))
				{
					return;
				}
				startRename();
			}
		});

		enabledBox.setSelected(cameraPoint.isEnabled());
		enabledBox.addItemListener(e -> {
			cameraPoint.setEnabled(enabledBox.isSelected());
			plugin.updateConfig();
		});

		namePanel.add(nameInput, BorderLayout.CENTER);
		namePanel.add(nameActions, BorderLayout.EAST);

		JPanel bottomPanel = new JPanel(new BorderLayout(0, 6));
		bottomPanel.setBackground(ColorScheme.DARKER_GRAY_COLOR);
		bottomPanel.setBorder(new EmptyBorder(8, 8, 8, 8));

		keybindButton.setToolTipText("Load camera point hotkey");
		keybindButton.setText(cameraPoint.getKeybind().toString());
		keybindButton.setFont(FontManager.getDefaultFont().deriveFont(12f));
		keybindButton.addMouseListener(new MouseAdapter()
		{
			@Override
			public void mouseReleased(MouseEvent e)
			{
				keybindButton.setText(Keybind.NOT_SET.toString());
				cameraPoint.setKeybind(Keybind.NOT_SET);
				plugin.updateConfig();
			}
		});
		keybindButton.addKeyListener(new KeyAdapter()
		{
			@Override
			public void keyPressed(KeyEvent e)
			{
				if (e.getKeyCode() == KeyEvent.VK_ESCAPE)
				{
					requestFocusInWindow();
					return;
				}
				if (Helper.isModifierKey(e.getKeyCode()))
				{
					return;
				}
				Keybind hotkey = new Keybind(e);
				keybindButton.setText(hotkey.toString());
				cameraPoint.setKeybind(hotkey);
				plugin.updateConfig();
				requestFocusInWindow();
			}
		});

		if (cameraPoint.getDirection() == null)
		{
			cameraPoint.setDirection(CameraPoint.Direction.NORTH);
		}
		directionDropdown.setSelectedItem(cameraPoint.getDirection());
		directionDropdown.setToolTipText("Direction to face when loading this point");
		directionDropdown.addActionListener(e -> {
			CameraPoint.Direction selectedDirection = (CameraPoint.Direction) directionDropdown.getSelectedItem();
			if (selectedDirection == null)
			{
				return;
			}
			cameraPoint.setDirection(selectedDirection);
			plugin.updateConfig();
		});

		applyZoomButton.setSelected(cameraPoint.isApplyZoom());
		applyZoomButton.setToolTipText("Apply saved zoom when loading this point");
		applyZoomButton.setIcon(zoomIcon);
		applyZoomButton.setSelectedIcon(zoomPressedIcon);
		applyZoomButton.setPressedIcon(zoomPressedIcon);
		applyZoomButton.setBorderPainted(false);
		applyZoomButton.setContentAreaFilled(false);
		applyZoomButton.setFocusPainted(false);
		applyZoomButton.setCursor(Cursor.getPredefinedCursor(Cursor.HAND_CURSOR));
		applyZoomButton.setPreferredSize(new Dimension(18, 18));
		applyZoomButton.addItemListener(e -> {
			cameraPoint.setApplyZoom(applyZoomButton.isSelected());
			plugin.updateConfig();
		});

		JPanel buttonsPanel = new JPanel(new BorderLayout(4, 0));
		buttonsPanel.setBackground(ColorScheme.DARKER_GRAY_COLOR);

		updateZoomLabel.setCursor(Cursor.getPredefinedCursor(Cursor.HAND_CURSOR));
		updateZoomLabel.setBackground(ColorScheme.DARKER_GRAY_COLOR);
		updateZoomLabel.setToolTipText("Update zoom from current game zoom");
		updateZoomLabel.setPreferredSize(new Dimension(18, 18));
		updateZoomLabel.addMouseListener(new MouseAdapter()
		{
			@Override
			public void mousePressed(MouseEvent e)
			{
				if (!Helper.checkClick(e))
				{
					return;
				}
				updateZoomLabel.setIcon(fromGamePressedIcon);
			}

			@Override
			public void mouseReleased(MouseEvent e)
			{
				if (!Helper.checkClick(e))
				{
					return;
				}
				int result = JOptionPane.showConfirmDialog(PointPanel.this,
					"Are you sure you want to update zoom from the game?",
					"Are you sure?",
					JOptionPane.OK_CANCEL_OPTION);
				if (result == 0)
				{
					plugin.updateCameraPointZoom(cameraPoint);
					applyZoomButton.setSelected(cameraPoint.isApplyZoom());
					plugin.updateConfig();
				}
				updateZoomLabel.setIcon(fromGameHoverIcon);
			}

			@Override
			public void mouseEntered(MouseEvent e)
			{
				updateZoomLabel.setIcon(fromGameHoverIcon);
			}

			@Override
			public void mouseExited(MouseEvent e)
			{
				updateZoomLabel.setIcon(fromGameIcon);
			}
		});

		deleteLabel.setCursor(Cursor.getPredefinedCursor(Cursor.HAND_CURSOR));
		deleteLabel.setBackground(ColorScheme.DARKER_GRAY_COLOR);
		deleteLabel.setToolTipText("Delete camera point");
		deleteLabel.setPreferredSize(new Dimension(18, 18));
		deleteLabel.addMouseListener(new MouseAdapter()
		{
			@Override
			public void mousePressed(MouseEvent e)
			{
				if (!Helper.checkClick(e))
				{
					return;
				}
				deleteLabel.setIcon(deletePressedIcon);
			}

			@Override
			public void mouseReleased(MouseEvent e)
			{
				if (!Helper.checkClick(e))
				{
					return;
				}
				int result = JOptionPane.showConfirmDialog(PointPanel.this,
					"Are you sure you want to delete this camera point?",
					"Are you sure?",
					JOptionPane.OK_CANCEL_OPTION);
				if (result == 0)
				{
					plugin.getCameraPointGroupManager().removePointFromGroup(group, cameraPoint);
					plugin.updateConfig();
					reloadFunction.run();
				}
				deleteLabel.setIcon(deleteHoverIcon);
			}

			@Override
			public void mouseEntered(MouseEvent e)
			{
				deleteLabel.setIcon(deleteHoverIcon);
			}

			@Override
			public void mouseExited(MouseEvent e)
			{
				deleteLabel.setIcon(deleteIcon);
			}
		});

		buttonsPanel.add(updateZoomLabel, BorderLayout.WEST);
		buttonsPanel.add(deleteLabel, BorderLayout.EAST);

		directionDropdown.setPreferredSize(new Dimension(100, 22));
		keybindButton.setPreferredSize(new Dimension(78, 22));

		JPanel topMetaRow = new JPanel(new FlowLayout(FlowLayout.LEFT, 6, 0));
		topMetaRow.setBackground(ColorScheme.DARKER_GRAY_COLOR);
		topMetaRow.add(enabledBox);
		topMetaRow.add(directionDropdown);

		JPanel bottomControlsRow = new JPanel(new FlowLayout(FlowLayout.LEFT, 6, 0));
		bottomControlsRow.setBackground(ColorScheme.DARKER_GRAY_COLOR);
		bottomControlsRow.add(keybindButton);
		bottomControlsRow.add(applyZoomButton);
		bottomControlsRow.add(updateZoomLabel);
		bottomControlsRow.add(deleteLabel);

        zoomSpinner.setToolTipText("Zoom value");
        zoomSpinner.setValue(cameraPoint.getZoom());
        zoomSpinner.setEnabled(cameraPoint.isApplyZoom());
        zoomSpinner.addChangeListener(e ->
        {
            cameraPoint.setZoom((int)zoomSpinner.getValue());
            plugin.updateConfig();
        });

        JPanel centerRow = new JPanel(new FlowLayout(FlowLayout.RIGHT, 0, 0));
        centerRow.setBackground(ColorScheme.DARKER_GRAY_COLOR);
        centerRow.add(zoomSpinner);

		bottomPanel.add(topMetaRow, BorderLayout.NORTH);
        bottomPanel.add(centerRow, BorderLayout.CENTER);
		bottomPanel.add(bottomControlsRow, BorderLayout.SOUTH);

		add(namePanel, BorderLayout.NORTH);
		add(bottomPanel, BorderLayout.SOUTH);
	}

	private void configureNameActions(JPanel nameActions)
	{
		saveLabel.setVisible(false);
		saveLabel.setFont(FontManager.getRunescapeSmallFont());
		saveLabel.setForeground(ColorScheme.PROGRESS_COMPLETE_COLOR);
		saveLabel.addMouseListener(new MouseAdapter()
		{
			@Override
			public void mousePressed(MouseEvent e)
			{
				if (!Helper.checkClick(e))
				{
					return;
				}
				saveLabel.setForeground(ColorScheme.PROGRESS_COMPLETE_COLOR.brighter());
			}

			@Override
			public void mouseReleased(MouseEvent e)
			{
				if (!Helper.checkClick(e))
				{
					return;
				}
				saveName();
				saveLabel.setForeground(ColorScheme.PROGRESS_COMPLETE_COLOR.darker());
			}

			@Override
			public void mouseEntered(MouseEvent e)
			{
				saveLabel.setForeground(ColorScheme.PROGRESS_COMPLETE_COLOR.darker());
			}

			@Override
			public void mouseExited(MouseEvent e)
			{
				saveLabel.setForeground(ColorScheme.PROGRESS_COMPLETE_COLOR);
			}
		});

		cancelLabel.setVisible(false);
		cancelLabel.setFont(FontManager.getRunescapeSmallFont());
		cancelLabel.setForeground(ColorScheme.PROGRESS_ERROR_COLOR);
		cancelLabel.addMouseListener(new MouseAdapter()
		{
			@Override
			public void mousePressed(MouseEvent e)
			{
				if (!Helper.checkClick(e))
				{
					return;
				}
				cancelLabel.setForeground(ColorScheme.PROGRESS_ERROR_COLOR.brighter());
			}

			@Override
			public void mouseReleased(MouseEvent e)
			{
				if (!Helper.checkClick(e))
				{
					return;
				}
				cancelRename();
				cancelLabel.setForeground(ColorScheme.PROGRESS_ERROR_COLOR.darker());
			}

			@Override
			public void mouseEntered(MouseEvent e)
			{
				cancelLabel.setForeground(ColorScheme.PROGRESS_ERROR_COLOR.darker());
			}

			@Override
			public void mouseExited(MouseEvent e)
			{
				cancelLabel.setForeground(ColorScheme.PROGRESS_ERROR_COLOR);
			}
		});

		renameLabel.setFont(FontManager.getRunescapeSmallFont());
		renameLabel.setForeground(ColorScheme.LIGHT_GRAY_COLOR.darker());
		renameLabel.addMouseListener(new MouseAdapter()
		{
			@Override
			public void mousePressed(MouseEvent e)
			{
				if (!Helper.checkClick(e))
				{
					return;
				}
				renameLabel.setForeground(ColorScheme.LIGHT_GRAY_COLOR);
			}

			@Override
			public void mouseReleased(MouseEvent e)
			{
				if (!Helper.checkClick(e))
				{
					return;
				}
				startRename();
				renameLabel.setForeground(ColorScheme.LIGHT_GRAY_COLOR.darker().darker());
			}

			@Override
			public void mouseEntered(MouseEvent e)
			{
				renameLabel.setForeground(ColorScheme.LIGHT_GRAY_COLOR.darker().darker());
			}

			@Override
			public void mouseExited(MouseEvent e)
			{
				renameLabel.setForeground(ColorScheme.LIGHT_GRAY_COLOR.darker());
			}
		});

		nameActions.add(saveLabel, BorderLayout.EAST);
		nameActions.add(cancelLabel, BorderLayout.WEST);
		nameActions.add(renameLabel, BorderLayout.CENTER);
	}

	private void startRename()
	{
		nameInput.setEditable(true);
		saveLabel.setVisible(true);
		cancelLabel.setVisible(true);
		renameLabel.setVisible(false);
		nameInput.getTextField().requestFocusInWindow();
		nameInput.getTextField().selectAll();
	}

	private void saveName()
	{
		nameInput.setEditable(false);
		cameraPoint.setName(nameInput.getText());
		plugin.updateConfig();
		saveLabel.setVisible(false);
		cancelLabel.setVisible(false);
		renameLabel.setVisible(true);
		requestFocusInWindow();
	}

	private void cancelRename()
	{
		nameInput.setEditable(false);
		nameInput.setText(cameraPoint.getName());
		saveLabel.setVisible(false);
		cancelLabel.setVisible(false);
		renameLabel.setVisible(true);
		requestFocusInWindow();
	}
}

File: src/main/java/com/camerapoints/ui/PointPanel.kt
package com.camerapoints.ui

import com.camerapoints.CameraPointsPlugin
import com.camerapoints.models.CameraPoint
import com.camerapoints.models.CameraPointGroup
import net.runelite.client.config.Keybind
import net.runelite.client.ui.ColorScheme
import net.runelite.client.ui.FontManager
import net.runelite.client.ui.components.FlatTextField
import net.runelite.client.util.ImageUtil
import java.awt.BorderLayout
import java.awt.Color
import java.awt.Cursor
import java.awt.Dimension
import java.awt.event.KeyAdapter
import java.awt.event.KeyEvent
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import javax.swing.*
import javax.swing.border.CompoundBorder
import javax.swing.border.EmptyBorder
import javax.swing.border.LineBorder
import javax.swing.border.MatteBorder

class PointPanel(
    private val plugin: CameraPointsPlugin,
    private val group: CameraPointGroup,
    private val cameraPoint: CameraPoint,
    private val reloadFunction: () -> Unit
) : JPanel() {

    val FROM_GAME_ICON = ImageIcon(ImageUtil.loadImageResource(CameraPointsPlugin::class.java, "from_game_icon.png"))
    val FROM_GAME_HOVER_ICON =
        ImageIcon(ImageUtil.loadImageResource(CameraPointsPlugin::class.java, "from_game_icon_blue.png"))
    val FROM_GAME_PRESSED_ICON = ImageIcon(ImageUtil.alphaOffset(FROM_GAME_HOVER_ICON.image, -50))

    val DELETE_ICON = ImageIcon(ImageUtil.loadImageResource(CameraPointsPlugin::class.java, "delete_icon.png"))
    val DELETE_HOVER_ICON =
        ImageIcon(ImageUtil.loadImageResource(CameraPointsPlugin::class.java, "delete_icon_red.png"))
    val DELETE_PRESSED_ICON = ImageIcon(ImageUtil.alphaOffset(DELETE_HOVER_ICON.image, -50))


    val enabledBox: JCheckBox = JCheckBox()
    val nameInput: FlatTextField = FlatTextField()
    val saveLabel = JLabel("Save")
    val cancelLabel = JLabel("Cancel")
    val renameLabel = JLabel("Rename")
    val keybindButton: JButton = JButton()
    val fromGameLabel: JLabel = JLabel(FROM_GAME_ICON)
    val deleteLabel: JLabel = JLabel(DELETE_ICON)

    init {
        layout = BorderLayout()
        border = BorderFactory.createLineBorder(Color.LIGHT_GRAY)
        background = ColorScheme.DARKER_GRAY_COLOR

        // Row 1: Checkbox and Name
        val namePanel = JPanel(BorderLayout())
        namePanel.background = ColorScheme.DARKER_GRAY_COLOR
        namePanel.border = CompoundBorder(
            MatteBorder(0, 0, 1, 0, ColorScheme.DARK_GRAY_COLOR),
            LineBorder(ColorScheme.DARKER_GRAY_COLOR)
        )

        val nameActions = JPanel(BorderLayout(4, 0))
        nameActions.background = ColorScheme.DARKER_GRAY_COLOR
        nameActions.border = EmptyBorder(0, 4, 0, 8)

        saveLabel.isVisible = false
        saveLabel.font = FontManager.getRunescapeSmallFont()
        saveLabel.foreground = ColorScheme.PROGRESS_COMPLETE_COLOR
        saveLabel.addMouseListener(object: MouseAdapter() {
            override fun mousePressed(e: MouseEvent) {
                if (!Helper.checkClick(e))
                    return
                saveLabel.foreground = ColorScheme.PROGRESS_COMPLETE_COLOR.brighter()
            }

            override fun mouseReleased(e: MouseEvent) {
                if (!Helper.checkClick(e))
                    return
                nameInput.setEditable(false)
                cameraPoint.name = nameInput.text
                plugin.updateConfig()
                saveLabel.isVisible = false
                cancelLabel.isVisible = false
                renameLabel.isVisible = true
                requestFocusInWindow()
                saveLabel.foreground = ColorScheme.PROGRESS_COMPLETE_COLOR.darker()
            }

            override fun mouseEntered(e: MouseEvent) {
                saveLabel.foreground = ColorScheme.PROGRESS_COMPLETE_COLOR.darker()
            }

            override fun mouseExited(e: MouseEvent) {
                saveLabel.foreground = ColorScheme.PROGRESS_COMPLETE_COLOR
            }
        })

        cancelLabel.isVisible = false
        cancelLabel.font = FontManager.getRunescapeSmallFont()
        cancelLabel.foreground = ColorScheme.PROGRESS_ERROR_COLOR
        cancelLabel.addMouseListener(object: MouseAdapter() {
            override fun mousePressed(e: MouseEvent) {
                if (!Helper.checkClick(e))
                    return
                cancelLabel.foreground = ColorScheme.PROGRESS_ERROR_COLOR.brighter()
            }

            override fun mouseReleased(e: MouseEvent) {
                if (!Helper.checkClick(e))
                    return
                nameInput.setEditable(false)
                nameInput.text = cameraPoint.name
                saveLabel.isVisible = false
                cancelLabel.isVisible = false
                renameLabel.isVisible = true
                requestFocusInWindow()
                cancelLabel.foreground = ColorScheme.PROGRESS_ERROR_COLOR.darker()
            }

            override fun mouseEntered(e: MouseEvent) {
                cancelLabel.foreground = ColorScheme.PROGRESS_ERROR_COLOR.darker()
            }

            override fun mouseExited(e: MouseEvent) {
                cancelLabel.foreground = ColorScheme.PROGRESS_ERROR_COLOR
            }
        })

        renameLabel.font = FontManager.getRunescapeSmallFont()
        renameLabel.foreground = ColorScheme.LIGHT_GRAY_COLOR.darker()
        renameLabel.addMouseListener(object: MouseAdapter() {
            override fun mousePressed(e: MouseEvent) {
                if (!Helper.checkClick(e))
                    return
                renameLabel.foreground = ColorScheme.LIGHT_GRAY_COLOR
            }

            override fun mouseReleased(e: MouseEvent) {
                if (!Helper.checkClick(e))
                    return
                nameInput.setEditable(true)
                saveLabel.isVisible = true
                cancelLabel.isVisible = true
                renameLabel.isVisible = false
                nameInput.textField.requestFocusInWindow()
                nameInput.textField.selectAll()
                renameLabel.foreground = ColorScheme.LIGHT_GRAY_COLOR.darker().darker()
            }

            override fun mouseEntered(e: MouseEvent) {
                renameLabel.foreground = ColorScheme.LIGHT_GRAY_COLOR.darker().darker()
            }

            override fun mouseExited(e: MouseEvent) {
                renameLabel.foreground = ColorScheme.LIGHT_GRAY_COLOR.darker()
            }
        })

        nameActions.add(saveLabel, BorderLayout.EAST)
        nameActions.add(cancelLabel, BorderLayout.WEST)
        nameActions.add(renameLabel, BorderLayout.CENTER)

        nameInput.text = cameraPoint.name
        nameInput.border = null
        nameInput.setEditable(false)
        nameInput.background = ColorScheme.DARKER_GRAY_COLOR
        nameInput.preferredSize = Dimension(0, 24)
        nameInput.textField.foreground = Color.WHITE
        nameInput.textField.border = EmptyBorder(0, 8, 0, 0)
        nameInput.addKeyListener(object: KeyAdapter() {
            override fun keyPressed(e: KeyEvent) {
                if (e.keyCode == KeyEvent.VK_ENTER) {
                    nameInput.setEditable(false)
                    cameraPoint.name = nameInput.text
                    plugin.updateConfig()
                    saveLabel.isVisible = false
                    cancelLabel.isVisible = false
                    renameLabel.isVisible = true
                    requestFocusInWindow()
                } else if (e.keyCode == KeyEvent.VK_ESCAPE) {
                    nameInput.setEditable(false)
                    nameInput.text = cameraPoint.name
                    saveLabel.isVisible = false
                    cancelLabel.isVisible = false
                    renameLabel.isVisible = true
                    requestFocusInWindow()
                }
            }
        })
        nameInput.textField.addMouseListener(object: MouseAdapter() {
            override fun mouseClicked(e: MouseEvent) {
                if (!Helper.checkClick(e))
                    return
                nameInput.setEditable(true)
                saveLabel.isVisible = true
                cancelLabel.isVisible = true
                renameLabel.isVisible = false
                nameInput.textField.requestFocusInWindow()
                nameInput.textField.selectAll()
            }
        })

        enabledBox.isSelected = cameraPoint.enabled
        enabledBox.addItemListener {
            cameraPoint.enabled = enabledBox.isSelected
            plugin.updateConfig()
        }

        namePanel.add(enabledBox, BorderLayout.WEST)
        namePanel.add(nameInput, BorderLayout.CENTER)
        namePanel.add(nameActions, BorderLayout.EAST)

        // Row 2: Button, Up Arrow, X
        val bottomPanel = JPanel(BorderLayout())
        bottomPanel.background = ColorScheme.DARKER_GRAY_COLOR
        bottomPanel.border = EmptyBorder(8, 8, 8, 8)

        keybindButton.toolTipText = "Load camera point hotkey"
        keybindButton.text = cameraPoint.keybind.toString()
        keybindButton.font = FontManager.getDefaultFont().deriveFont(12f)
        keybindButton.addMouseListener(object : MouseAdapter() {
            override fun mouseReleased(e: MouseEvent) {
                keybindButton.text = Keybind.NOT_SET.toString()
                cameraPoint.keybind = Keybind.NOT_SET
                plugin.updateConfig()
            }
        })
        keybindButton.addKeyListener(object : KeyAdapter() {
            override fun keyPressed(e: KeyEvent) {
                if (e.keyCode == KeyEvent.VK_ESCAPE) {
                    requestFocusInWindow()
                    return
                }
                val hotkey = Keybind(e)
                keybindButton.text = hotkey.toString()
                cameraPoint.keybind = hotkey
                plugin.updateConfig()
                requestFocusInWindow()
            }
        })

        val buttonsPanel = JPanel(BorderLayout(4, 0))
        buttonsPanel.background = ColorScheme.DARKER_GRAY_COLOR

        fromGameLabel.cursor = Cursor.getPredefinedCursor(Cursor.HAND_CURSOR)
        fromGameLabel.background = ColorScheme.DARKER_GRAY_COLOR
        fromGameLabel.toolTipText = "Get camera details from game"
        fromGameLabel.addMouseListener(object : MouseAdapter() {
            override fun mousePressed(e: MouseEvent) {
                if (!Helper.checkClick(e))
                    return
                fromGameLabel.icon = FROM_GAME_PRESSED_ICON
            }

            override fun mouseReleased(e: MouseEvent) {
                if (!Helper.checkClick(e))
                    return
                val result = JOptionPane.showConfirmDialog(this@PointPanel, "Are you sure you want to load the camera details from the game?", "Are you sure?", JOptionPane.OK_CANCEL_OPTION)
                if (result == 0) {
                    plugin.saveCameraPosition(cameraPoint)
                    plugin.updateConfig()
                }
                fromGameLabel.icon = FROM_GAME_HOVER_ICON
            }

            override fun mouseEntered(e: MouseEvent?) {
                fromGameLabel.icon = FROM_GAME_HOVER_ICON
            }

            override fun mouseExited(e: MouseEvent?) {
                fromGameLabel.icon = FROM_GAME_ICON
            }
        })

        deleteLabel.cursor = Cursor.getPredefinedCursor(Cursor.HAND_CURSOR)
        deleteLabel.background = ColorScheme.DARKER_GRAY_COLOR
        deleteLabel.toolTipText = "Delete camera point"
        deleteLabel.addMouseListener(object : MouseAdapter() {
            override fun mousePressed(e: MouseEvent) {
                if (!Helper.checkClick(e))
                    return
                deleteLabel.icon = DELETE_PRESSED_ICON
            }

            override fun mouseReleased(e: MouseEvent) {
                if (!Helper.checkClick(e))
                    return
                val result = JOptionPane.showConfirmDialog(this@PointPanel, "Are you sure you want to delete this camera point?", "Are you sure?", JOptionPane.OK_CANCEL_OPTION)
                if (result == 0) {
                    plugin.cameraPointGroupManager.removePointFromGroup(group, cameraPoint)
                    plugin.updateConfig()
                    reloadFunction()
                }
                deleteLabel.icon = DELETE_HOVER_ICON
            }

            override fun mouseEntered(e: MouseEvent?) {
                deleteLabel.icon = DELETE_HOVER_ICON
            }

            override fun mouseExited(e: MouseEvent?) {
                deleteLabel.icon = DELETE_ICON
            }
        })

        buttonsPanel.add(fromGameLabel, BorderLayout.WEST)
        buttonsPanel.add(deleteLabel, BorderLayout.EAST)

        bottomPanel.add(keybindButton, BorderLayout.WEST)
        bottomPanel.add(buttonsPanel, BorderLayout.EAST)

        add(namePanel, BorderLayout.NORTH)
        add(bottomPanel, BorderLayout.SOUTH)
    }
}